### PR TITLE
WIP: Add a better example for file upload

### PR DESCRIPTION
**Pull request by barrachri**
## What do these changes do?

Improve docs.

I think most of the time when you upload a file you also have related fields.

This PR add a general case to the docs., where you have a text field and file field.

## Are there changes in behavior for the user?
No.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `changes` folder
  * name it `<issue_id>.<type>` for example (588.bug)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by asvetlov** [comment]
Hmm. The code is definitely robust but the snippet goes to very low details maybe.
@kxepal please review.

**Comment by barrachri** [comment]
Thanks @asvetlov.

Do you think is too verbose or more comments/explanations outside of the code would improve this?

**Comment by fafhrd91** [comment]
I’d add note in file operations, in general case file operations block

File: docs/web.rst [hunk 1]
```diff
@@ -737,28 +742,49 @@ that :meth:`Request.post` reads the whole payload in memory, resulting in possib
...
 <aiohttp-multipart>`::
 
-    async def store_mp3_handler(request):
-
-        reader = await request.multipart()
-
-        # /!\ Don't forget to validate your inputs /!\
+    async def store_multipart_handler(request):
+        """Handle a multipart form."""
+        if request.method == 'POST':
```
Comment (kxepal) [inline docs/web.rst]: I think this is a bad pattern to recommend. You should have this condition on router level.

File: docs/web.rst [hunk 2]
```diff
@@ -737,28 +742,49 @@ that :meth:`Request.post` reads the whole payload in memory, resulting in possib
...
+                            content_type.startswith('text/'):
+                        charset = field.get_charset(default='utf-8')
+                        value = value.decode(charset)
+                    data[field.name] = value
+                    size += len(value)
+
+            return web.Response(text='Thanks {} for uploading {}'
+                                .format(data['name'], data['file']))
+
+        return web.Response(body=page, content_type="text/html")
```
Comment (kxepal) [inline docs/web.rst]: page is not defined.

**Comment by kxepal** [comment]
@fafhrd91 
Good point. However, that's not the only edge case for this example.

**Comment by barrachri** [comment]
So:

* Add a note that file operations generally block.

@kxepal a small note in the official docs completed with a link to a good example in `/examples`  would be better?

**Comment by asvetlov** [comment]
I think we need the docs change but it should look like:
```python
    async def store_mp3_handler(request):

        reader = await request.multipart()

        # /!\ Don't forget to validate your inputs /!\

        field = await reader.next()
        assert field.name == 'name'
        name = await field.read(decode=True)

        field = await reader.next()
        assert field.name == 'mp3'
        filename = field.filename
        # You cannot rely on Content-Length if transfer is chunked.
        size = 0
        with open(os.path.join('/spool/yarrr-media/mp3/', filename), 'wb') as f:
            while True:
                chunk = await field.read_chunk()  # 8192 bytes by default.
                if not chunk:
                    break
                size += len(chunk)
                f.write(chunk)

        return web.Response(text='{} sized of {} successfully stored'
                                 ''.format(filename, size))
```

The key is using `field.name` and `field.read()`.
Maybe we need process `field.get_charset()` too.

@kxepal does it make sense?

**Comment by kxepal** [comment]
@barrachri 
Yes.

@asvetlov 
That's nice. I believe you would like to use `name`  instead of `filename` on file open line. Otherwise it's unused and may confuse.

**Comment by asvetlov** [comment]
I have no strong opinion.
My intention is showing all useful field methods and attributes

**Comment by barrachri** [comment]
The current example and @asvetlov's are very clear in showing how you can chunk files.

The problem is that usually in other frameworks you don't loop or iterate on an object to get your params and the same is also true for aiohttp, with `Request.query` and `data = await request.post()`.

My problem was that you jump from this example

```python
    mp3 = data['mp3']

    # .filename contains the name of the file in string format.
    filename = mp3.filename

    # .file contains the actual file data that needs to be stored somewhere.
    mp3_file = data['mp3'].file

    content = mp3_file.read()
```

to this example

```python
async def store_mp3_handler(request):

    reader = await request.multipart()

    # /!\ Don't forget to validate your inputs /!\

    mp3 = await reader.next()

    filename = mp3.filename

    # You cannot rely on Content-Length if transfer is chunked.
    size = 0
    with open(os.path.join('/spool/yarrr-media/mp3/', filename), 'wb') as f:
        while True:
            chunk = await mp3.read_chunk()  # 8192 bytes by default.
            if not chunk:
                break
            size += len(chunk)
            f.write(chunk)

    return web.Response(text='{} sized of {} successfully stored'
                             ''.format(filename, size))
```
and you completely miss what's going on with `next` because there isn't a context anymore.
The short paragraph before the example talk about the multipart reader and if you follow the [link](http://aiohttp.readthedocs.io/en/stable/multipart.html#aiohttp-multipart) everything is discussed apart from `Reading Multipart Request`.

After I checked the implementation of `Request.post` everything was clear but I agree is too verbose to just paste the code.

I think that a simple comment will make the example more clear.
Something like this:

```python
    async def store_mp3_handler(request):

        reader = await request.multipart()

        # /!\ Don't forget to validate your inputs /!\

        # reader.next() will `yield` the fields of your form

        field = await reader.next()
        assert field.name == 'name'
        name = await field.read(decode=True)

        field = await reader.next()
        assert field.name == 'mp3'
        filename = field.filename
        # You cannot rely on Content-Length if transfer is chunked.
        size = 0
        with open(os.path.join('/spool/yarrr-media/mp3/', filename), 'wb') as f:
            while True:
                chunk = await field.read_chunk()  # 8192 bytes by default.
                if not chunk:
                    break
                size += len(chunk)
                f.write(chunk)

        return web.Response(text='{} sized of {} successfully stored'
                                 ''.format(filename, size))
```

**Comment by asvetlov** [comment]
I think now is better. Please update the RP.
@kxepal ?

**Comment by barrachri** [comment]
Shall I add also an example in `examples/` or just updating the PR?

**Comment by kxepal** [comment]
Let's update `examples/` in another PR. It's not a blocker for docs improvement.

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**